File: src/main/scala/resonant/lib/grid/NodeGrid.java
package resonant.lib.grid;

import java.util.AbstractMap;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;

import net.minecraftforge.common.ForgeDirection;

public abstract class NodeGrid<N extends Node> extends Grid<N>
{
    public NodeGrid(Class<? extends N> type)
    {
        super(type);
    }

    @Override
    protected void reconstructNode(N node)
    {
        node.recache();
        node.setGrid(this);

        AbstractMap<Object, ForgeDirection> connections = new HashMap(node.getConnections());

        for (Object connection : connections.keySet())
        {
            if (isValidNode(connection) && connection instanceof Node)
            {
                Node connectedNode = (Node) connection;

                if (connectedNode.getGrid() != this)
                {
                    synchronized (connectedNode.getGrid().nodes)
                    {
                        connectedNode.getGrid().nodes.clear();
                    }

                    add((N) connectedNode);
                    reconstructNode((N) connectedNode);
                }
            }
        }
    }

    @Override
    public void deconstruct()
    {
        synchronized (nodes)
        {
            Iterator<N> it = new HashSet<N>(nodes).iterator();

            while (it.hasNext())
            {
                N node = it.next();
                node.setGrid(null);
                node.reconstruct();
            }

            nodes.clear();
        }
    }
}

File: src/main/scala/resonant/lib/grid/node/NodeGrid.scala
package resonant.lib.grid.node

import resonant.api.grid.INodeProvider
import resonant.lib.grid.GridNode

/**
 * A node that is part of a larger grid of nodes
 *
 * @tparam A The type this node connects to
 * @author Calclavia
 */
abstract class NodeGrid[A <: NodeGrid[A]](parent: INodeProvider) extends NodeConnector[A](parent) //with INodeGrid
{
  protected var _grid: GridNode[A] = null

  /** Event called after a grid reconstructs itself */
  var onGridReconstruct: () => Unit = () => ()

  override def reconstruct()
  {
    super.reconstruct()
    grid.reconstruct(this.asInstanceOf[A])
  }

  override def deconstruct()
  {
    super.deconstruct()
    grid.deconstruct(this.asInstanceOf[A])
  }

  def setGrid(grid: GridNode[_])
  {
    this._grid = grid.asInstanceOf[GridNode[A]]
  }

  def grid: GridNode[A] =
  {
    if (_grid == null)
      _grid = newGrid

    _grid
  }

  protected def newGrid: GridNode[A]
}
